### PR TITLE
Update 1.1.6

### PR DESCRIPTION
**Pull request by expressmailing**
Main update : The subscription is made just after the installation. On the configuration page.
Language correction for tinymce
Default subscriber selection is executed on the selection step
Email step 3 : the view is split into 3 tabs (HTML editor, File import, Url import)
Deletion of the local page "filters_buy". The page is always fetched from the API
Improvement of customer selections
Improvement for multishop
Selection of newsletter subscribers

### CODE REVIEW

**Comment by gfanton** [comment]
Hello @expressmailing ,

Just few points:
-  When you switch your store in English and you want to open a new account: "Terms of Use" is in French.
- Also i got this error when i try to open a new account: "Error during communication with Express-Mailing API : GEN_Non_Implemente ". Did i something wrong?

mailexpress_open_account

Regards,
Guilhem

**Comment by expressmailing** [comment]
Hello @gfanton,

* For now, we don't have the Terms of Use translated in English. It will be translated soon. And it will appear automaticaly on the module.

* Yesterday we had an issue with our API. This is working now so if you tried yesterday, it's normal.

Thanks

Regards,
Thierry

**Comment by expressmailing** [comment]
Hello @gfanton ,

There is a new update (the last commit) that i've add to the pull request. This will correct an issue for one of our customers. 

Is it required to have the "Terms of Use" page translated before you merge the pull request ?

Thanks

Regards,
Thierry

**Comment by gfanton** [comment]
Hello @expressmailing ,

It's okay if the the "Terms of Use"  is updated automatically.
Otherwise everything looks good.

Regards,

@Quetzacoalt91 You can merge the PR :+1:

File: controllers/admin/adminmarketingestep3.php [hunk 1]
```diff
@@ -366,6 +398,8 @@ private function getHTMLContentDB()
 			if ($pos = strpos($domain_name, ':'))
 				$domain_name = Tools::substr($domain_name, 0, $pos);
 
+			$this->context->smarty->assign('shop_name', Shop::getShops()[Shop::getCurrentShop()]['name']);
```
Comment (Quetzacoalt91) [inline controllers/admin/adminmarketingestep3.php]: Be careful, this line won't be compatible with PHP 5.2.
You have to store the result of `Shop::getShops()` first, and then use it with `[Shop::getCurrentShop()]['name']`

File: controllers/admin/adminmarketingestep3.php [hunk 2]
```diff
@@ -624,11 +665,35 @@ private function cleanHTML()
...
+		}
+
+		$links = $this->getLinkTags($this->html_content);
+		$hrefs = array_unique($links[2]);
+		foreach ($hrefs as $key => $href)
+		{
+			$corrected_link = implode('/', array_map('rawurlencode', explode('/', $href)));
+			$corrected_link = str_replace('%3A//', '://', $corrected_link);
+			$corrected_link = str_replace('%23%23', '##', $corrected_link);
+			$this->html_content = str_replace($links[0][$key], $links[1][$key].$corrected_link.$links[3][$key], $this->html_content);
```
Comment (Quetzacoalt91) [inline controllers/admin/adminmarketingestep3.php]: You have a common code for your loops. You execute the same loop on a variable which can be huge. Here is a lighter version which seems to fit your needs (not tested):

```php
	private function encodeURLs()
	{
		$elements = array_merge($this->getImgTags($this->html_content), $this->getLinkTags($this->html_content));
		$srcs = array_unique($elements[2]);
		foreach ($srcs as $key => $src)
		{
			$corrected_src = implode('/', array_map('rawurlencode', explode('/', $src)));
			$corrected_src = str_replace(array('%3A//', '%23%23'), array('://', '##'), $corrected_src);
			$this->html_content = str_replace($elements[0][$key], $elements[1][$key].$corrected_src.$elements[3][$key], $this->html_content);
		}
	}
```

With this version, you spare 1 loop execution and 1 call to `str_replace`.

With this version, you should gain some execution time :)

File: controllers/admin/db_marketing.php [hunk 3]
```diff
@@ -135,30 +140,54 @@ public static function getCustomersEmailRequest($campaign_id,
...
+		$checked_products, $checked_categories, $checked_shops,
 		$paying_filters = array(), $extended = false,
 		$limit = 0, &$list_total = null)
 	{
-		$req = selft::getCustomersEmailRequest($campaign_id,
+		$req = self::getCustomersEmailRequest($campaign_id,
 												$checked_langs, $checked_groups,
 												$checked_campaign_optin, $checked_campaign_newsletter, $checked_campaign_active,
-												$checked_products, $checked_categories,
+												$checked_products, $checked_categories, $checked_shops,
```
Comment (Quetzacoalt91) [inline controllers/admin/db_marketing.php]: Can you check that you did not forget the parameter `$checked_campaign_guest` ?

File: upgrade/Upgrade-1.1.6.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * http://www.express-mailing.com for more information.
+ *
+ * @author    Axalone France <info@express-mailing.com>
+ * @copyright 2014-2015 (c) Axalone France
+ * @license   http://opensource.org/licenses/GPL-3.0  GNU General Public License, version 3 (GPL-3.0)
+ */
+
+function upgrade_module_1_1_6($module)
+{
+	$return = Db::getInstance()->execute('ALTER TABLE `ps_expressmailing_email`
```
Comment (Quetzacoalt91) [inline upgrade/Upgrade-1.1.6.php]: Do not hardcode `ps_` in your SQL request, but use `_DB_PREFIX_` instead !

File: upgrade/Upgrade-1.1.6.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+		CHANGE COLUMN `campaign_optin` `campaign_optin` ENUM(\'1\',\'0\') NOT NULL DEFAULT \'0\' AFTER `campaign_selected_recipients`,
+		CHANGE COLUMN `campaign_newsletter` `campaign_newsletter` ENUM(\'1\',\'0\') NOT NULL DEFAULT \'0\' AFTER `campaign_optin`,
+		ADD COLUMN `campaign_guest` ENUM(\'1\',\'0\') NOT NULL DEFAULT \'0\' AFTER `campaign_active`;');
+
+	$return &= Db::getInstance()->execute('ALTER TABLE `ps_expressmailing_email_recipients`
```
Comment (Quetzacoalt91) [inline upgrade/Upgrade-1.1.6.php]: Same comment about hardcoded `ps_`

**Comment by Quetzacoalt91** [comment]
Hi @expressmailing,

I'm reviewing the code before merging your PR, and a SQL table is missing in your upgrade file. You should create the table `expressmailing_email_shops_groups` in it, don't you ?

File: translations/fr.php [hunk 6]
```diff
@@ -1,9 +1,10 @@
 <?php
 
 global $_MODULE;
+global $test;
```
Comment (Quetzacoalt91) [inline translations/fr.php]: This global can be removed.

File: views/templates/admin/configuration.tpl [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	</div>
+</div>
+<div class="col-lg-4">
+	<div class="em-fixed-col-lg">
+		<div class="row row-margin-bottom intro-left-blocs">
+			<div class="col-lg-12 text-center">
+				<h2 class="text-uppercase" style="margin-top: 0px">{l s='Subscription' mod='expressmailing'}</h2>
+				<a href="index.php?controller=AdminMarketingInscription&token={Tools::getAdminTokenLite('AdminMarketingInscription')|escape:'html':'UTF-8'}" class="btn btn-primary btn-lg" >{l s='Open your account' mod='expressmailing'}</a>
+				<div style="padding: 1em">
+					<div style="font-size: 15px; margin-right: auto; margin-left: auto; display: inline; text-align: left">
```
Comment (Quetzacoalt91) [inline views/templates/admin/configuration.tpl]: In all this file, you should really think to move the style definition in a CSS file.

**Comment by expressmailing** [comment]
Hello @Quetzacoalt91 ,

Thanks a lot, everything has been corrected now.

**Comment by Quetzacoalt91** [comment]
Hi @expressmailing,

Thanks, your PR has just been merged and the changes will be available from tomorrow.

Best regards

**Pull request merged**